python/pyspark/sql/types.py

```diff
@@ -1108,19 +1109,23 @@ def _has_nulltype(dt):
         return isinstance(dt, NullType)
 
 
-def _merge_type(a, b):
+def _merge_type(a, b, path=''):
```

Comment:
Yup, I agree with it for the current status .. 

it's kind of difficult to come up with a good message format in such cases. To me, I actually kind of gave up a pretty format and just chose prose before (in the PR #18521).

I still don't have a good idea to show the nested structure in the error message to be honest. I was thinking one of prose, kind of piece of codes (like `schema['f1'].dataType.elementType.keyType` instead of https://github.com/apache/spark/pull/19792#issuecomment-346681841), or somehow pretty one like `printSchema()` ... ?

Maybe, I am thinking of referring other formats in Spark or somewhere like Pandas or piece of codes for now.

So, to cut it short, here are what are on my mind for the example of https://github.com/apache/spark/pull/19792#issuecomment-346681841:

1. Prose

```
TypeError: key in map field in array element in field f1: Can not blabla
```

2. Piece of codes with `StructType`

```
TypeError: schema(?)['f1'].dataType.elementType.keyType: Can not blabla
TypeError: struct(?)['f1'].dataType.elementType.keyType: Can not blabla
```


3. `printSchema()`

```
TypeError: 
root
 |-- f1: array (nullable = false)
 |    |-- element: map (containsNull = false)
 |    |    |-- key: integer*
 |    |    |-- value: long (valueContainsNull = false)
: *Can not blabla
```

Reply:
I really don't have a strong opinion on this. @ueshin do you maybe have a preference or another option, or are you okay with the current format in general?

Reply:
Maybe, just printing the string in SQL could be an option too to be consistent .. it's a bit messy for deeply nested schema though.:

```
TypeError: struct<f1:array<map<int,string>>> and struct<f1:array<map<int,int>>> Can not blabla
```

Reply:
I prefer the format we can know the path to the error and I guess we don't need the pretty one like the `printSchema()`.
Maybe following the #18521 format is good enough, and the current one is good as well but I wanted something in front of the path string instead of starting with `.`.

Reply:
Yup, looks we need something ahead anyway. To me, I am okay with any format. 

I think we could fix later again. Just I wondered if you have a preference.

Reply:
Hi folks, 

Great options proposed by @HyukjinKwon - though I did not comprehend what was the conclusion of the discussion with @ueshin :)

Which format should we employ? And do you want me to use this format right now or is it something you'll fix later?

Thanks!

Reply:
Well, can you follow the #18521 format for now? Thanks!